### PR TITLE
Remove mapping type

### LINKED ISSUE #301: Tutorial computation of accuracy and recall is wrong
It makes an implicit assumption that there is a 100% overlap.
The problem seems to be in the `describe_accuracy` method.

### PR DESCRIPTION
**Pull request by wilko77**
The entity service changed the supported result types. 'mapping' is a thing of the past now.
This PR adjusts the tutorials and the cli to this new reality. 

also closes #301 as I changed the way we compute recall and precision

### CODE REVIEW

File: docs/tutorial_api.ipynb [hunk 1]
```diff
@@ -599,7 +600,7 @@
     "    )\n",
     "    solution = anonlink.solving.greedy_solve(results_candidate_pairs)\n",
     "    print('Found {} matches'.format(len(solution)))\n",
-    "    return {a:b for ((_, a),(_, b)) in solution}"
+    "    return set((a, b) for ((_, a), (_, b)) in solution)"
```
Comment (gusmith) [inline docs/tutorial_api.ipynb]: May be good to show here that we didn't have any duplicate (as not expected), but wrking with a set will help later to compute the accuracy and recall.

File: docs/tutorial_api.ipynb [hunk 2]
```diff
@@ -642,45 +643,29 @@
    },
    "outputs": [],
    "source": [
-    "a_csv.seek(0)\n",
-    "b_csv.seek(0)\n",
-    "a_raw = a_csv.readlines()\n",
-    "b_raw = b_csv.readlines()\n",
+    "dfA_ = dfA.rename(lambda x: x[4:-4], axis='index').reset_index()\n",
```
Comment (gusmith) [inline docs/tutorial_api.ipynb]: That would be good in the tutorial to describe why you are transforming the index values.

File: docs/tutorial_api.ipynb [hunk 3]
```diff
@@ -642,45 +643,29 @@
...
-    "    print(\"We've got {} true positives, {} false positives, and {} false negatives.\".format(TP, FP, FN))\n",
-    "    print('Precision: {:.3f}, Recall: {:.3f}, Accuracy: {:.3f}'.format(\n",
-    "        TP/(TP+FP), \n",
-    "        TP/(TP+FN), \n",
-    "        (TP+TN)/(TP+TN+FP+FN)))\n",
-    " \n"
+    "    precision = tp / (tp + fp)\n",
+    "    recall = tp / (tp + fn)\n",
+    "\n",
+    "    print('Precision: {:.2f}, Recall: {:.2f}'.format(precision, recall))"
```
Comment (gusmith) [inline docs/tutorial_api.ipynb]: Any reason behind removing the accuracy? And if we definitely remove the accuracy, we should rename the method which is currently named `describe_accuracy`

File: docs/tutorial_api.ipynb [hunk 4]
```diff
@@ -762,10 +744,17 @@
    "source": [
     "Great, for this threshold value we get a precision of 100% and a recall of 99.2%. \n",
```
Comment (gusmith) [inline docs/tutorial_api.ipynb]: Will need to change these values based on the one you are computing. With this run, recall is 99%, not 99.2.

File: docs/tutorial_api.ipynb [hunk 5]
```diff
@@ -762,10 +744,17 @@
...
     "\n",
-    "The explanation is that when the information about an entity differs slightly in the two datasets (e.g. spelling errors, abbrevations, missing values, ...) then the corresponding CLKs will differ in some number of bits as well. It is important to choose an appropriate threshold for the amount of perturbations present in the data (a threshold of 0.72 and below generates a perfect mapping without mistakes).\n",
+    "The explanation is that when the information about an entity differs slightly in the two datasets (e.g. spelling errors, abbrevations, missing values, ...) then the corresponding CLKs will differ in some number of bits as well. It is important to choose an appropriate threshold for the amount of perturbations present in the data (a threshold of 0.72 and below generates an almost perfect mapping with little mistakes).\n",
```
Comment (gusmith) [inline docs/tutorial_api.ipynb]: :+1:

File: docs/tutorial_cli.ipynb [hunk 6]
```diff
@@ -412,9 +406,7 @@
    "source": [
     "## Hash the data\n",
     "\n",
-    "We can now hash our Personally Identifiable Information (PII) data from the CSV file using our defined linkage schema. We must provide one *secret* to this command - this secret has to be used by both parties hashing data. For this toy example we will use the secret _'secret'_, for real data, make sure that the secret contains enough entropy, as knowledge of this secret is sufficient to reconstruct the PII information from a CLK!\n",
-    "\n",
-    "Also, **do not share this secret with anyone, except the other participating party.**"
+    "We can now hash our Personally Identifiable Information (PII) data from the CSV file using our defined linkage schema. We must provide two *secret keys* to this command - these keys have to be used by both parties hashing data. For this toy example we will use the keys _'key1'_ and _'key2'_, for real data, make sure that the keys contain enough entropy, as knowledge of these keys is sufficient to reconstruct the PII information from a CLK! Also, **do not share these keys with anyone, except the other participating party.**"
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: That's an old text still mentioning using two keys... I'm not sure what happened there to get back the old one...

File: docs/tutorial_cli.ipynb [hunk 7]
```diff
@@ -523,17 +511,15 @@
...
+    "pycharm": {}
    },
    "outputs": [
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
       "{\n",
-      "  \"version\": 2,\n",
+      "  \"version\": 3,\n",
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: :+1:

File: docs/tutorial_cli.ipynb [hunk 8]
```diff
@@ -858,7 +842,7 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "{\"project_count\": 762, \"rate\": 382537, \"status\": \"ok\"}\r\n"
+      "{\"project_count\": 9, \"rate\": 19071855, \"status\": \"ok\"}\r\n"
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: That was a fresh entity-service :)

File: docs/tutorial_cli.ipynb [hunk 9]
```diff
@@ -1128,65 +1102,49 @@
     "pycharm": {}
    },
    "source": [
-    "Let's investigate some of those matches and the overall matching quality. In this case we have the ground truth so we can compute the precision, recall, and accuracy."
+    "Let's investigate some of those matches and the overall matching quality. In this case we have the ground truth so we can compute the precision and recall."
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: So definitely removing the accuracy from the output.

File: docs/tutorial_cli.ipynb [hunk 10]
```diff
@@ -1115,11 +1086,14 @@
     }
    ],
    "source": [
-    "with open(\"results.txt\") as f:\n",
-    "    str_mapping = json.load(f)['mapping']\n",
+    "def extract_matches(file):\n",
+    "    with open(file, 'rt') as f:\n",
+    "        results = json.load(f)['groups']\n",
+    "    found_matches = set((a, b) for ((_, a), (_, b)) in results)\n",
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: Same comment a on the other tutorial about the set and having no duplicates.

File: docs/tutorial_cli.ipynb [hunk 11]
```diff
@@ -1128,65 +1102,49 @@
...
+    "dfB_ = dfB.rename(lambda x: x[4:-6], axis='index').reset_index()\n",
+    "a = pd.DataFrame({'ida': dfA_.index, 'link': dfA_['rec_id']})\n",
+    "b = pd.DataFrame({'idb': dfB_.index, 'link': dfB_['rec_id']})\n",
+    "dfj = a.merge(b, on='link', how='inner').drop(columns=['link'])\n",
+    "true_matches = set((row[0], row[1]) for row in dfj.itertuples(index=False))\n",
     "\n",
-    "num_entities = len(b_raw) - 1\n",
     "\n",
-    "def describe_accuracy(mapping, show_examples=False):\n",
+    "def describe_accuracy(found_matches, show_examples=False):\n",
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: Same comment about renaming the method if not providing accuracy.

File: docs/tutorial_cli.ipynb [hunk 12]
```diff
@@ -1128,65 +1102,49 @@
...
-    }
+    "pycharm": {}
    },
    "outputs": [],
    "source": [
-    "with open(\"PII_a.csv\", 'rt') as f:\n",
-    "    a_raw = f.readlines()\n",
-    "with open(\"PII_b.csv\", 'rt') as f:\n",
-    "    b_raw = f.readlines()\n",
+    "dfA_ = dfA.rename(lambda x: x[4:-4], axis='index').reset_index()\n",
```
Comment (gusmith) [inline docs/tutorial_cli.ipynb]: same comment explaining what we are doing here.

File: docs/tutorial_api.ipynb (hunk 3 above)
Comment (wilko77) [reply docs/tutorial_api.ipynb]: good point.

**Pull request merged**